src/core/ddsc/src/dds_rhc_default.c

```diff
@@ -1016,22 +1019,30 @@ static int inst_accepts_sample_by_writer_guid (const struct rhc_instance *inst,
 
 static int inst_accepts_sample (const struct dds_rhc_default *rhc, const struct rhc_instance *inst, const struct ddsi_writer_info *wrinfo, const struct ddsi_serdata *sample, const bool has_data)
 {
-  if (rhc->by_source_ordering)
```

Comment:
I think the "time based filter" QoS should also apply when "by reception timestamp" order is selected - that is at least what I would expect if I imagine myself to be a regular user. This way it only does it for "by source timestamp". That is, in principle, trivially dealt with.

Another thing is that C says signed integer over/underflow is undefined behaviour, and that needs to be accounted for when checking whether the sample arrived in the "minimum separation" windor or not.

IINM, then, on input, the source time stamp of the sample and source timestamp stored in the instance can be anything (unavoidably so: data received from other implementations must be assumed to can be anything, and if the instance is created based on an sample with an "invalid" timestamp, then the only possible value for the instance is also "invalid").

If either (or both) of the timestamps is "invalid", then there is no meaningful time difference. I am tempted to say it means they are unordered with respect to each other. There is no way of building an eventually consistent system based on unordered samples, so we probably should also require that a "by source timestamp" writer *must* provide an non-"invalid" timestamp < ∞ for each eample, and discard everything else.

Then, we can observe that in the "by source timestamp" case, they are always ordered (even if the arithmetic is a bit tricky). For "by reception timestamp" we recognise they may be unordered, in which case we simply consider the new one to be sufficiently more recent. Alternatively, we could use another bit of information - e.g., the local clock - but there are dragons there too. Smart people just avoid these problems, use Cyclone DDS and are always guaranteed to have valid source timestamps 🙂

So I am wondering if it doesn't make the most sense to write it like:
```
if (by_source_ordering) { // (essentially) unchanged
  if (sample->timestamp.v > inst->tstamp.v)
    ;//possibly accept
  else if (sample->timestamp.v < inst->tstamp.v)
    return 0;// reject
  else if (!inst_accepts_sample_by_writer_guid (...))
    return 0;//reject
}
if (minimum_separation > 0 && sample->timestamp.v != INVALID && inst->tstamp.v != INVALID) {
  // new code; we can actually compute the separation because we made sure we have a
  // non-"invalid" timestamps
  if (sample->timestamp.v < INT64_MIN + minimum_separation || sample->timestamp.v - minimum_separation < inst->tstamp)
    return 0;//reject
  }
}
if (rhc->exclusive_ownership && inst->wr_iid_islive && inst->wr_iid != wrinfo->iid)
...
```

Which really only adds a bit for `minimum_separation` and the arithmetic is done in a way that avoids trouble. What do you think?

Reply:
Having a `by_reception_timestamp` kind of filter will require some larger modifications, so we maybe do that in another PR and now just keep unsupported?

Maybe have some warnings or something when these "unsupported" or incompatible filter settings are attempted?

Reply:
The fact that a time-based filter is basically defined (in the spec) as a means to limit the number of 'seen updates' of arrived samples of instances, its (at least to me) something that's about 'local reader-side' behavior and when that's combined with the fact that most users use _by_reception_timestamp_ ordering (as source-timestamp-ordering requires time-alignment throughout the system), I'm very much in-favor of allowing time-based filtering for that (common) use-case... just wondering how that implies a 'large/larger modification'  ..

Reply:
The larger modification is defining/recording the reception timestamp, because at the moment it is not stored in either the `rhc_instance` or the `ddsi_serdata` entities.
And this information is needed to keep track of whether to store/drop samples in the reader.

Reply:
I think there's another possible view: that is that "destination order" switch really is (roughly) about whether something describes events you want to queue (`by_reception_timestamp`) or state you want to update (`by_source_timestamp`); and that it has nothing whatsoever to do with how you interpret "time-based filter".

That view allows you to use the source timestamp for the "time-based filter" in both cases. Of course it falls apart when people start using the source timestamp in creative ways, but the answer is simple: then don't do that.

In Cyclone (and in OpenSplice) there is *always* a source timestamp, it is never elided because someone chose "by_reception_timestamp" for the "destination_order" setting in the writer. That's one of the silliest things I have seen done ...

So in my view, the only problem is the existence of some *other* implementations where a "by reception timestamp" writer may not provide a valid timestamp. I'd just accept everything and suggest to use a proper implementation for the writing side as well to anyone who runs into that detail and considers it a problem.

Reply:
If we are going for the option to not implement time based filters based on reception timestamps, which is the status quo at this moment in time, I do think that giving a warning when you attempt to use time filtering when the ordering is by_reception_timestamp is giving at least the correct feedback to the user.
We can then later choose to implement a reception timestamp functionality.